File: api-spring-boot/src/main/java/com/example/api/entity/Etudiant.java
package com.example.api.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.Period;

import com.fasterxml.jackson.annotation.JsonFormat;

@Entity
@Table(name = "etudiants")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Etudiant {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String cin;

    private String nom;

    @JsonFormat(pattern = "yyyy-MM-dd")   // ✅ Annotation déplacée ici
    private LocalDate dateNaissance;

    // ✅ Constructeur personnalisé (sans id, car généré automatiquement)
    public Etudiant(String cin, String nom, LocalDate dateNaissance) {
        this.cin = cin;
        this.nom = nom;
        this.dateNaissance = dateNaissance;
    }

    // Méthode utilitaire pour calculer l'âge
    public int age() {
        return Period.between(this.dateNaissance, LocalDate.now()).getYears();
    }
}

File: grading-service/mobile-app/lib/models/etudiant.dart
class Etudiant {
  final int id;
  final String cin;
  final String nom;
  final String dateNaissance;

  Etudiant({
    required this.id,
    required this.cin,
    required this.nom,
    required this.dateNaissance,
  });

  /// Parse un objet JSON reçu de l'API Spring Boot.
  /// Jackson est configuré pour sérialiser LocalDate en "YYYY-MM-DD".
  factory Etudiant.fromJson(Map<String, dynamic> json) {
    return Etudiant(
      id: json['id'] as int,
      cin: json['cin'] as String? ?? '',
      nom: json['nom'] as String? ?? '',
      dateNaissance: json['dateNaissance'] as String? ?? '',
    );
  }

  @override
  String toString() => 'Etudiant(id: $id, cin: $cin, nom: $nom)';
}
